File: solutions/go/03/deque_window/main.go
package main

import "fmt"

func maxSumK(xs []int, k int) int {
	s, ans := 0, -1<<60
	left := 0
	for right, x := range xs {
		s += x
		if right-left+1 > k {
			s -= xs[left]
			left++
		}
		if right-left+1 == k {
			if s > ans {
				ans = s
			}
		}
	}
	return ans
}

func main() {
	xs := []int{1, 2, 3, 4, 5}
	fmt.Println(maxSumK(xs, 3))
}

File: solutions/python/03/deque_window/main.py
from collections import deque


def max_sum_k(xs: list, k: int) -> int:
    q = deque()
    s = 0
    ans = -10**18
    for i, x in enumerate(xs):
        s += x
        q.append(i)
        if len(q) > k:
            s -= xs[q.popleft()]
        if i >= k - 1:
            ans = max(ans, s)
    return ans


def main() -> None:
    xs = [1, 2, 3, 4, 5]
    print(max_sum_k(xs, 3))


if __name__ == "__main__":
    main()
